dt_parse_iso.c

```diff
@@ -432,61 +432,135 @@ dt_parse_iso_zone_lenient(const char *str, size_t len, int *op) {
+ * Allows both ISO8601 and extended Tarantool datetime formats
+ */
+static size_t
+count_dashes(const unsigned char *p, size_t i, size_t len) {
+    size_t n = 0;
+
+    for (; i < len; i++) {
+        switch (p[i]) {
+        case '0': case '1': case '2': case '3': case '4':
+        case '5': case '6': case '7': case '8': case '9':
+        case 'Q': case 'W':
+            continue;
+        case '-':
+            n++;
+            continue;
+        }
+        break;
+    }
+    return n;
+}
+#endif
 /*
  *  Basic      Extended
  *  20121224   2012-12-24   Calendar date   (ISO 8601)
  *  2012359    2012-359     Ordinal date    (ISO 8601)
  *  2012W521   2012-W52-1   Week date       (ISO 8601)
  *  2012Q485   2012-Q4-85   Quarter date
+ *
+ *  Tarantool extended ranges
+ * #ifdef DT_PARSE_ISO_TNT
+ *  -001-12-31      0000-01-01
+ *  -5879610-06-22  5879611-07-11
+ * #endif
  */
 size_t
 dt_parse_iso_date(const char *str, size_t len, dt_t *dtp) {
-    const unsigned char *p;
+    const unsigned char *p = (const unsigned char *)str;
     int y, x, d;
     size_t n;
     dt_t dt;
+    int sign = +1;
+    int head_n = 4;
+#ifdef DT_PARSE_ISO_TNT
+    int dashes_n;
+
+    if (p[0] == '-') {
+        sign = -1;
+        p++;
+        len--;
+    }
+    dashes_n = count_dashes(p, 0, len);
+#endif
 
-    p = (const unsigned char *)str;
     n = count_digits(p, 0, len);
     switch (n) {
+#ifdef DT_PARSE_ISO_TNT
+        case 3: /* -001 (year) */
+        case 4: /* -2001 (year) */
+            if (dashes_n == 2) {
+                y = parse_number(p, 0, n);
+                break;
+            } else if (n == 4) {  /* 2012 (year) */
+                y = parse_number(p, 0, 4);
+                break;
+            } else {
+                return 0;
+            }
+        case 5: /* 10000-01-01 (extended Tarantol range) */
+        case 6: /* 109000-01-01 (extended Tarantol range) */
+        case 7: /* 5879611-07-11 (extended Tarantol range) */
+            if (dashes_n == 2) {
+                y = parse_number(p, 0, n);
+                head_n = n;
+                break;
+            } else if (n == 7) { /* 2012359 (basic ordinal date) */
+                y = parse_number(p, 0, 4);
+                d = parse_number(p, 4, 3);
+                p += 7;
+                goto yd;
+            } else {
+                return 0;
+            }
+#else
         case 4: /* 2012 (year) */
             y = parse_number(p, 0, 4);
             break;
         case 7: /* 2012359 (basic ordinal date) */
             y = parse_number(p, 0, 4);
             d = parse_number(p, 4, 3);
+            p += 7;
             goto yd;
+#endif
         case 8: /* 20121224 (basic calendar date) */
             y = parse_number(p, 0, 4);
             x = parse_number(p, 4, 2);
             d = parse_number(p, 6, 2);
+            p += 8;
             goto ymd;
         default:
             return 0;
     }
 
-    if (len < 8)
+    if (len < (head_n + 4))
```

Comment:
Let me clarify it a bit:
- we want to execute relatively the same logics, after we have parsed year part;
- but what used to be done only for 4 characters long year, we now do for 3..7 chars long;
- thus all these constants no reduced by 4 and added head_n (i.e. instead of `8 we have `4 + head_n`)